### ISSUE TITLE
Incorrect v-site positions

### ISSUE THREAD

**Issue by SimonBoothroyd** [openforcefield/openff-interchange#832]
**Description**

V-site positions seem to be incorrectly calculated as of 0.3.16. I've only tested divalent lone pairs, but this may also apply to other types.

**Reproduction**

```
import numpy
import openff.interchange
import openff.toolkit
import openff.units
import openmm.unit


def compute_v_site_coords(
    conformer: numpy.ndarray, interchange: openff.interchange.Interchange
) -> numpy.ndarray:
    n_v_sites = len(interchange.collections["VirtualSites"].key_map)

    conformer = (
        numpy.vstack([conformer, numpy.zeros((n_v_sites, 3))]) * openmm.unit.angstrom
    )

    openmm_system = interchange.to_openmm()
    openmm_context = openmm.Context(
        openmm_system,
        openmm.VerletIntegrator(0.1),
        openmm.Platform.getPlatformByName("Reference"),
    )
    openmm_context.setPositions(conformer)
    openmm_context.computeVirtualSites()
    conformer_with_v_sites = openmm_context.getState(getPositions=True).getPositions(
        asNumpy=True
    )

    v_site_coords = conformer_with_v_sites[(len(conformer) - n_v_sites) :, :]
    return v_site_coords.value_in_unit(openmm.unit.angstrom)


def compute_out_of_plane_angle(
    parent_coord, orient_coord_1, orient_coord_2, v_site_coord
):
    normal = numpy.cross(orient_coord_1 - parent_coord, orient_coord_2 - parent_coord)
    angle_with_normal = numpy.arccos(
        numpy.dot(v_site_coord, normal)
        / (numpy.linalg.norm(v_site_coord) * numpy.linalg.norm(normal))
    )
    angle_with_plane = numpy.pi / 2 - angle_with_normal

    return numpy.rad2deg(angle_with_plane)


def main():
    conformer = numpy.array([[0.0, 0.0, 0.0], [-1.0, -1.0, 0.0], [1.0, -1.0, 0.0]])

    molecule = openff.toolkit.Molecule.from_mapped_smiles("[H:2][O:1][H:3]")
    molecule._conformers = [conformer * openff.units.unit.angstrom]

    interchange = openff.interchange.Interchange.from_smirnoff(
        openff.toolkit.ForceField("tip5p.offxml"),
        molecule.to_topology(),
    )

    v_site_coords = compute_v_site_coords(conformer, interchange)

    print("v-site coords")
    print(v_site_coords)

    expected_distance = 0.7  # angstrom
    expected_angle = 54.735  # degrees

    dist_1 = numpy.linalg.norm(v_site_coords[0, :] - conformer[0, :])
    dist_2 = numpy.linalg.norm(v_site_coords[1, :] - conformer[0, :])

    print(f"expected dist={expected_distance} dist1={dist_1} dist2={dist_2}")

    angle_1 = compute_out_of_plane_angle(
        conformer[0, :], conformer[1, :], conformer[2, :], v_site_coords[0, :]
    )
    angle_2 = compute_out_of_plane_angle(
        conformer[0, :], conformer[1, :], conformer[2, :], v_site_coords[1, :]
    )

    print(f"expected angle={expected_angle} angle1={angle_1} angle2={angle_2}")


if __name__ == "__main__":
    main()
```

**Output**

For 0.3.15:

```
v-site coords
[[ 0.          0.40415128 -0.5715433 ]
 [ 0.          0.40415128  0.5715433 ]]
expected dist=0.7 dist1=0.7000000000000001 dist2=0.7000000000000001
expected angle=54.735 angle1=-54.73500000000001 angle2=54.735
```

for 0.3.16:

```
v-site coords
[[ 0.          0.68981653 -1.28875807]
 [ 0.          0.68981653 -1.28875807]]
expected dist=0.7 dist1=1.4617606513224397 dist2=1.4617606513224397
expected angle=54.735 angle1=-61.84178049146057 angle2=-61.84178049146057
```

Here the distance and OOP angle are incorrect and the v-sites both end up in the same location.

Please include the output, including full tracebacks of any errors, resulting from the reproduction.

**Software versions**

- Which operating system and version are you using?
- How did you install Interchange?
- What is the output of running `conda list`?

**Comment by mattwthompson** [comment]
That's disappointing, I apologize for the regression. I thought matching the weights of [OpenMM's implementation](https://github.com/openmm/openmm/blob/8.0.0/wrappers/python/openmm/app/data/tip5p.xml#L14-L15) would produce accurate geometries. 

I noticed that fiddling with the input conformers produces different results; I think this is because the math used to determine the weights uses the geometry specified by the force field (here, H-O distance and H-O-H angle), not what's found in the input conformer, so mixing the two will produce funky business. 

Conceptually I don't know how to work around this as SMIRNOFF defines virtual site positions in real space and OpenMM defines them as a weighted average of particle positions, so either the force field geometry would need to be used or the weights re-computed whenever the geometry of the molecule changes (every timestep). `LocalCoordinatesSite` did not have the issue, for reasons that have always escaped me.

Debugging below

----

Everything else from an `openmm.XmlSerializer.serialize` in the middle of your code passes the eye check:

* Particles ordered O H H V V
* Virtual sites defined by atoms O H1 H2 and O H2 H1 - though I see here that maybe ordering should not be backwards _and_ $w_{cross}$ have opposite sign, which wouldn't explain all of what you're seeing
* Every intra-atomic pair is excluded
* Constraints look to have normal TIP-like values

```xml
<?xml version="1.0" ?>
<System openmmVersion="8.0" type="System" version="1">
	<PeriodicBoxVectors>
		<A x="2" y="0" z="0"/>
		<B x="0" y="2" z="0"/>
		<C x="0" y="0" z="2"/>
	</PeriodicBoxVectors>
	<Particles>
		<Particle mass="15.99943"/>
		<Particle mass="1.007947"/>
		<Particle mass="1.007947"/>
		<Particle mass="0">
			<OutOfPlaneSite p1="0" p2="1" p3="2" w12="-.3449082628693362" w13="-.3449082628693362" wc="-6.443790349189734"/>
		</Particle>
		<Particle mass="0">
			<OutOfPlaneSite p1="0" p2="2" p3="1" w12="-.3449082628693362" w13="-.3449082628693362" wc="6.443790349189734"/>
		</Particle>
	</Particles>
	<Constraints>
		<Constraint d=".09572" p1="0" p2="1"/>
		<Constraint d=".09572" p1="0" p2="2"/>
		<Constraint d=".15139006545" p1="1" p2="2"/>
	</Constraints>
	<Forces>
		<Force alpha="0" cutoff=".9" dispersionCorrection="1" ewaldTolerance=".0005" exceptionsUsePeriodic="0" forceGroup="0" includeDirectSpace="1" ljAlpha="0" ljnx="0" ljny="0" ljnz="0" method="0" name="Nonbonded force" nx="0" ny="0" nz="0" recipForceGroup="-1" rfDielectric="78.3" switchingDistance=".8" type="NonbondedForce" useSwitchingFunction="1" version="4">
			<GlobalParameters/>
			<ParticleOffsets/>
			<ExceptionOffsets/>
			<Particles>
				<Particle eps=".66944" q="0" sig=".312"/>
				<Particle eps="0" q=".241" sig="1"/>
				<Particle eps="0" q=".241" sig="1"/>
				<Particle eps="0" q="-.241" sig="1"/>
				<Particle eps="0" q="-.241" sig="1"/>
			</Particles>
			<Exceptions>
				<Exception eps="0" p1="0" p2="1" q="0" sig="1"/>
				<Exception eps="0" p1="0" p2="2" q="0" sig="1"/>
				<Exception eps="0" p1="1" p2="2" q="0" sig="1"/>
				<Exception eps="0" p1="0" p2="3" q="0" sig="1"/>
				<Exception eps="0" p1="0" p2="4" q="0" sig="1"/>
				<Exception eps="0" p1="3" p2="1" q="0" sig="1"/>
				<Exception eps="0" p1="4" p2="1" q="0" sig="1"/>
				<Exception eps="0" p1="3" p2="2" q="0" sig="1"/>
				<Exception eps="0" p1="4" p2="2" q="0" sig="1"/>
				<Exception eps="0" p1="3" p2="4" q="0" sig="1"/>
			</Exceptions>
		</Force>
	</Forces>
</System>
```

**Comment by SimonBoothroyd** [comment]
🤷‍♂️ v-sites are hard and these things happen, no worries!

> LocalCoordinatesSite did not have the issue, for reasons that have always escaped me.

I think the main reason is that the `LocalCoordinatesSite` essentially builds an orthonormal basis (the key here I think being the *normal*), and then places the v-site relative to that basis using 'real space' positions (i.e. the exact distance, in plane, and out of plane angles).

I document the math [a little bit here](https://docs.openforcefield.org/projects/recharge/en/latest/users/theory.html#generating-coordinates), but the gist is I think they are the only way to represent SMIRNOFF defined v-sites given that, like you say, we define geometry in terms on distances and angles rather than relative coords!

**Comment by mattwthompson** [comment]
Ah, I was worried about that conclusion but didn't fully investigate the details. I'm really not fond of how human-illegible `LocalCoordinatesSite` is. I'll have to revert a bunch of changes, but maybe the API points (starting with getting positions of virtual sites in one method call, perhaps over time adding others over time based on feedback) can make it ergonomical enough. Crow for dinner tonight!

**Comment by mattwthompson** [comment]
@SimonBoothroyd I suppose you've already done the work of generating virtual site positions using only/mostly OpenFF objects?

The code may have changed (I think @lilyminium brought it into the new toolkit/Interchange APIs) but it looks like what I'm after here: https://github.com/openforcefield/openff-recharge/blob/0.5.0/openff/recharge/charges/vsite.py#L733

**Comment by SimonBoothroyd** [comment]
That should be it! A slightly more compact implementation here also: https://github.com/SimonBoothroyd/smee/blob/891460a85d6c2cea3adb3d2bbb1874d83caa094b/smee/geometry.py#L277

**Comment by mattwthompson** [comment]
After #835 this test case is back to working:

```
v-site coords
[[ 0.          0.40415128 -0.5715433 ]
 [ 0.          0.40415128  0.5715433 ]]
expected dist=0.7 dist1=0.7000000000000001 dist2=0.7000000000000001
expected angle=54.735 angle1=-54.73500000000001 angle2=54.735
```

I want to hammer away at a few tests to make sure the behavior is not limited to waters, and then I'll make a release

**Comment by mattwthompson** [comment]
With 0.3.17 this should be fixed, possibly modulo something going on with `MonovalentLonePairVirtualSite`s. Of course let me know if I botch it again 🏅
